### DISCUSSION TITLE
add support for general purpose middleware

### DISCUSSION THREAD

**Discussion by dominiquegarmier** [strawberry-graphql/strawberry#1890] [Ideas]
so a question was asked in discord about how to best reuse logic between fields (or queries and mutations). If the logic doesn't depend on arguments passed to the field we would use permission classes for this.

However if the logic depends on the inputs to that field there is currently, to my knowledge, no better way, than to outsource that logic into a function and call it at the beginning of each field. So i though about this and came up with this idea:

suppose we allowed an argument such as perhaps `middleware` to `strawberry.field`, which would be a list of functions, that get executed (in order), on the input. Forwarding the input (to the next middleware function or the field itself) if everything is fine, perhaps raising exceptions or even returning a response early. Importantly the input arguments should not be mutated as to allow for a predictable execution.

Perhaps a user would be able to define middleware by inheriting from some middleware class, where the user would have to implement a specific class method.
```python
from strawberry.middleware import BaseMiddleware, MiddlewareResponse

@strawberry.type
class Result:
    message: str

@strawberry.input
class Input:
    password: str
    ...

class MyMiddleware(BaseMiddleware):
    def run(self, info: Info, input: Input) -> MiddlewareResponse[Result]:
        user = info.context['user']
        if is_lucky_user(user):
            return MiddlewareResponse(Result(message="foo"))  # this will get returned no other middleware will run
        elif user.password != input.password:
            raise ValueError("wrong password")  # an exception will get returned
        return  # nothing happens, next middleware or field will run normally

@strawberry.type
class Query
    @strawberry.field(middleware=[MyMiddleware])
    async def my_field(self, input: Input) -> Result: ...
```

`run` would have to be able to handle all arguments provided to `my_field`, which i think should be fine. But at the moment I mainly see the problem that the middleware has to return a specific `Result` type.

What are your thoughts on this? would this be a useful feature? Is there already a better way to do this?

**Comment by LeKSuS-04** [comment]
I was the one asking the original question on discord.

As I see it, `run` method of the middleware should have return type that fits query/mutation. So, with your example, it would look like this:
```py
from strawberry.middleware import BaseMiddleware, MiddlewareResponse

@strawberry.type
class Result:
    message: str

@strawberry.type
class PossibleMiddlewareResult:
    something: str

@strawberry.type
class WrongPasswordError:
    details: str = 'Wrong password was provided'

@strawberry.input
class Input:
    password: str
    ...

MyQueryResult = strawberry.union(
    'MyQueryResult', (Result, PossibleMiddlewareResult, WrongPasswordError)
)

class MyMiddleware(BaseMiddleware):
    def run(self, info: Info, input: Input) -> MiddlewareResponse[PossibleMiddlewareResult, WrongPasswordError]:
        user = info.context['user']
        if is_lucky_user(user):
            return PossibleMiddlewareResultt(something="foo")  # this will get returned no other middleware will run
        elif user.password != input.password:
            return WrongPasswordError("wrong password")  # an exception will get returned
        return  # nothing happens, next middleware or field will run normally

@strawberry.type
class Query
    @strawberry.field(middleware=[MyMiddleware])
    async def my_field(self, input: Input) -> MyQueryResult: ...
```
Here `MyMiddleware.run` returns `PossibleMiddleWareResult` or `WrongPasswordError`. These types are also part of `MyQueryResult` union, which allows us to use that middleware on `my_field`. If types are different, strawberry should raise a runtime exception. `MiddlewareResponse` here works almost like `Optional`, by providing possibility that nothing will be returned (in that case, next middleware or field will run, as @DominiqueGarmier suggested)

I really like idea of being able to chain middlewares together. However, I'm not sure on how user will define such chains. Using just a list seems a bit dirty to me. But maybe that's me overthinking it because I can't find any real reason to why that would be inconvenient.

Also I've replaced raising `ValueError`, since I don't like the whole idea of raising an exception. That causes error to go to `errors` section of request. But as described [here](https://strawberry.rocks/docs/guides/errors#expected-errors), it's better to put at least some of the errors to `data` section to be able to provide error details to the user. That creates inconsistency, where some of the errors are displayed in one section and other are in another. So I tend to put all of the "expected" errors to `data` section. Unfortunately, `PermissionErrors` are placed into `errors` section and to my knowledge there's no workaround to change that behavior. If middlewares were to be implemented, that could completely replace permissions. That will help to solve that inconsistency. 

Overall, middlewares just seem like "better permissions" with some useful features, so in my opinion they would be quite a useful addition.

**Comment by dominiquegarmier** [comment]
So I thought about this some more. I think it would be nice if the response of such a middleware wouldn't be strictly tied to a single response type. Instead implement it as a dict internally, and allow any return types that will allow `Response(**mw_response)`. If you have a bunch of return types that all have an attribute `errors`, you could share middleware between them as long as the middleware returns a dict `{"errors": ...}`.

Here is some code I came up with.

```python
from typing import TypeVar
from typing import Generic
from typing import Type

import strawberry


K = TypeVar('K')


# the generic type indicates to what type
# we can coerce the response to be in _get_resp
class MiddlewareResponse(Generic[K]):
    def __init__(self, **kwargs):
        self.vars = {k: v for k, v in kwargs.items()}

    def _get_resp(self, cls: Type[K]) -> K:
        return cls(**self.vars)


T = TypeVar('T')


class BaseMiddleware(Generic[T]):

    _ret_cls: Type[T]

    def __init__(self, return_cls: Type[T]) -> None:
        self._ret_cls = return_cls

    # this method takes **kwargs so you are flexible with
    # how you want to use args
    # however it should return responses that can be coerced
    # to be of type T, as explained above
    def run(self, **kwargs) -> MiddlewareResponse[T] | None:
        raise NotImplementedError()


class MyMiddleware(BaseMiddleware):
    def run(self, **kwargs) -> MiddlewareResponse | None:
        if True:
            return MiddlewareResponse(foo='bar')  # return early
        elif False:
            raise Exception('foo')  # raise exception
        else:
            return None  # do nothing, run next


# this logic would live in strawberry.field, ...
def middleware_runner(input, field, middleware_list: list[BaseMiddleware]):
    for middleware in middleware_list:
        try:
            mw_resp = middleware.run()
        except Exception as e:  # perhaps logging?
            raise
        if isinstance(mw_resp, MiddlewareResponse):
            return mw_resp._get_resp(middleware._ret_cls)

    return field(input)


@strawberry.input
class Input:
    ...


@strawberry.field
class Response:
    foo: str = 'foo'
    bar: str = 'bar'


# we would now define middleware instances like so
mw_list = [MyMiddleware(Response)]
```

Im not sure if mypy will understand that `MyMiddleware(Response)` has type `MyMiddleware[Response]` and I'm also not sure if there are better ways to use generics here.

as for what you said about Exceptions. Yes I agree you probably shouldn't raise exceptions in middleware, its better to return graphql error types. But that's a larger question for strawberry, how to handle errors.

**Reply by dominiquegarmier** [reply]
I feel that fundamentally the issue is that there needs to be a uniform way to handle errors.

The easiest way is to just raise an Exception, which will eventually get caught and turned into a response. Bad for the client, because its hard to parse those responses.

The next best thing Is to always return Error Types.

@LeKSuS-04 , above you did that with `Union` I usually do that with another type:

```
@strawberry.type
FooResponse:
    success: bool
    data: ...
    errors: list[...]
```

crucially there are many ways to do this. From what I understand, they are all just as good. So wouldn't it be cool to be able to define that format somewhere? And then our middleware could respect that format?

**Reply by LeKSuS-04** [reply]
That's true, a universal way to handle errors would be nice. Your way of processing errors seems to duplicate GraphQL structure, which might be confusing. But again, that's just my opinion. Is there anything in GraphQL docs that describes handling complex errors, where more than just a message must be returned? I've tried searching and found nothing. So I am not sure if that's possible using just GraphQL, but strawberry definitely doesn't provide any convenient mechanisms to control what goes to `errors`, other then just raising python exceptions. 

After all, it logically seems that errors should be in `errors` section, not `data`. We just need a way to send more details in there.

**Reply by dominiquegarmier** [reply]
I modeled my graphql errors after this https://productionreadygraphql.com/2020-08-01-guide-to-graphql-errors .
I don't know what graphql spec has to say about this.

putting errors into `errors` robs us of type-safety though. I generally  tend to put unexpected errors there, and expected errors into `data`

But i think the best would be to let users define this however they want. So perhaps it would be nice to allow a wrapper around fields. Which could be achieved at the moment with decorators

**Reply by LeKSuS-04** [reply]
According to [this stackoverflow answer](https://stackoverflow.com/questions/37665539/mixing-of-schema-level-and-app-level-errors-in-graphql/37667798#37667798) and [this blog post](https://koistya.medium.com/validation-and-user-errors-in-graphql-mutations-39ca79cd00bf#.b3uc9juue), there's really no answer for that in GraphQL specification. So it's up to developers to decide, how exactly to implement error handling.

**Reply by dominiquegarmier** [reply]
ok, so in that case we would have to implement field-wrappers (decorators?)? if we want to give full control to the user?

there are two problems with that:

1) too many decorators may be hard to read / not sure if it's very pythonic? (I'm not really decided on that point)
 
    possible solution: we could accept the wrappers as a `list` as mentioned above and then execute them in order?

2) correctly type annotating decorators is not straightforward / strawberry will have to be able to correctly infer the types

    possible solution: create some decorators from which to inherit from?

**Reply by LeKSuS-04** [reply]
What if middleware were made to act almost like decorators? Not sure on implementation details, but for user it would look like this:
```py
@strawberry.type
class Query
    my_field: MyType = strawberry.field(my_resolver, middleware=[FirstMiddleware, SecondMiddleware])
```
... works as ...
```py
class Query:
    my_field: MyType = @strawberry.field(FirstMiddleware.run(SecondMiddleware.run(my_resolver)))
```
Where `MyType` encapsulates all possible outcomes.

**Reply by dominiquegarmier** [reply]
yes exactly, I would suggest that we would offer both apis, decorator and as a list. Since internally it would be built the same way. Similar to how you can use strawberry.field as a decorator and as a method.

**Reply by dominiquegarmier** [reply]
what im currently struggling with is:

say we have some fields:
```python
def foo() -> int: ...
def bar() -> str: ...
```

and we would like our middleware to return a generic `Response[T, V]` such that:
```python
def middleware(func: Callable[..., T]) -> Callable[..., Response[T]]: ...
```
then `middleware(foo)` and `middleware(bar)` should have the corresponding return types `Response[int]` and `Response[str]`. But at the moment I'm unable to forward these annotations properly. I would have to look at inspecting and setting the types at runtime. I have no idea how that works.

**Reply by dominiquegarmier** [reply]
the reason I'm bringing this up is, it would be nice if we could build generic middleware.

I was for example imagining an `ErrorHandler` middleware which could get instantiated with the response and expected error types `mw = ErrorHandler(InnerReponseType, [ErrorOne, ErrorTwo])`. Where `InnerResponseType`, `ErrorOne` and `ErrorTwo` are `strawberry.type`'s.

And now with a field `def field(...) -> InnerResponseType` and middleware as above, this would get turned into having a return type of `Response[InnerResponse, Union[ErrorOne, ErrorTwo]]` or something like that.

What `Response[...]` is defined as and how the conversion from `field()` to `Response[...]` happens would be left to the user to define.


EDIT:
i just found this:
https://github.com/strawberry-graphql/strawberry/blob/fae03c5836f3971364c92a90fd9b52ac23186e6c/strawberry/types/fields/resolver.py#L30

**Comment by matheushent** [comment]
has there been any progress in this discussion so far? I think there a bunch of use cases where the proposed middleware would be useful.

**Comment by matheushent** [comment]
Just brainstorming, wouldn't be possible to use [Field Extensions](https://strawberry.rocks/docs/guides/field-extensions#field-extensions) to handle the error logic outside of the resolver?
